librz/bin/format/le/le.c

```diff
@@ -57,828 +87,1749 @@ const char *le_get_arch(rz_bin_le_obj_t *bin) {
 	}
-	(*offset)++;
-	char *str = calloc((ut64)size + 1, sizeof(char));
-	rz_buf_read_at(buf, *offset, (ut8 *)str, size);
-	*offset += size;
-	return str;
+	RzBinImport *import = RZ_NEW0(RzBinImport);
+	if (!import) {
+	fail_cleanup:
+		rz_bin_import_free(import);
+		return NULL;
+	}
+	const char *libname = "";
+	if (le_imp->mod_ord - 1 < rz_pvector_len(bin->imp_mod_names)) {
+		libname = rz_pvector_at(bin->imp_mod_names, le_imp->mod_ord - 1);
+	}
+	if (le_imp->proc_name) {
+		CHECK(import->name = rz_str_newf("%s_%s", libname, le_imp->proc_name));
+	} else {
+		CHECK(import->name = rz_str_newf("%s_%u", libname, le_imp->proc_ord));
+	}
+	import->bind = RZ_BIN_BIND_GLOBAL_STR;
+	import->type = RZ_BIN_TYPE_UNKNOWN_STR;
+	CHECK(rz_list_append(bin->imports, import));
+	import->ordinal = ++bin->reloc_targets_count;
+	return import;
 }
 
-static RzBinSymbol *le_get_symbol(rz_bin_le_obj_t *bin, ut64 *offset) {
+static RZ_BORROW RzBinSymbol *le_add_symbol(rz_bin_le_obj_t *bin, ut32 ordinal, ut32 vaddr) {
 	RzBinSymbol *sym = RZ_NEW0(RzBinSymbol);
 	if (!sym) {
 		return NULL;
 	}
-	char *name = le_read_nonnull_str_at(bin->buf, offset);
-	if (!name) {
-		rz_bin_symbol_free(sym);
-		return NULL;
+	if (!ordinal) {
+		if (rz_list_empty(bin->symbols)) {
+			ordinal = 1;
+		} else {
+			ordinal = ((RzBinSymbol *)rz_list_tail(bin->symbols)->data)->ordinal + 1;
+		}
 	}
-	sym->name = name;
-	ut16 entry_idx;
-	if (!rz_buf_read_le16_offset(bin->buf, offset, &entry_idx)) {
+	if (!rz_list_append(bin->symbols, sym)) {
 		rz_bin_symbol_free(sym);
 		return NULL;
 	}
-	sym->ordinal = entry_idx;
+	sym->ordinal = ordinal;
+	sym->vaddr = vaddr;
+	sym->paddr = le_vaddr_to_paddr(bin, vaddr);
+	sym->bind = RZ_BIN_BIND_GLOBAL_STR;
+	sym->type = RZ_BIN_TYPE_UNKNOWN_STR;
 	return sym;
 }
 
-static bool read_le_entry_bundle_entry(RzBuffer *buf, ut64 addr, LE_entry_bundle_entry *e, LE_entry_bundle_type type) {
-	ut64 offset = addr;
-	switch (type) {
-	case ENTRY16:
-		return rz_buf_read8_offset(buf, &offset, &e->entry_16.flags) &&
-			rz_buf_read_le16_offset(buf, &offset, &e->entry_16.offset);
-	case CALLGATE:
-		return rz_buf_read8_offset(buf, &offset, &e->callgate.flags) &&
-			rz_buf_read_le16_offset(buf, &offset, &e->callgate.offset) &&
-			rz_buf_read_le16_offset(buf, &offset, &e->callgate.callgate_sel);
-	case ENTRY32:
-		return rz_buf_read8_offset(buf, &offset, &e->entry_32.flags) &&
-			rz_buf_read_le32_offset(buf, &offset, &e->entry_32.offset);
-	case FORWARDER:
-		return rz_buf_read8_offset(buf, &offset, &e->forwarder.flags) &&
-			rz_buf_read_le16_offset(buf, &offset, &e->forwarder.import_ord) &&
-			rz_buf_read_le32_offset(buf, &offset, &e->forwarder.offset);
-	default:
-		memset(e, 0, sizeof(LE_entry_bundle_entry));
+static RZ_BORROW LE_import *le_add_import(rz_bin_le_obj_t *bin,
+	ut32 mod_ord, bool proc_by_ord, ut32 proc, ut32 sym_ord) {
+
+	RzBinImport *bin_imp = NULL;
+	LE_import *le_imp = NULL;
+	char *proc_name = NULL;
+	if (false) {
+	fail_cleanup:
```

Comment:
The whole block unnecessary?

Reply:
same, used by CHECK macros, such blocks are all over the code and used the same way

Reply:
but it has `if (false)`

Reply:
Because this block is not entered except by goto. Such cleanup block is like group of "destructors", it follows declarations of variables to be destroyed in case of error (goto shouldn't cross initialization boundary). When those variables are initialized with e.g. ```x = malloc()```, there is a risk to enter cleanup block right away and then we have meaningful ```if (!x) {```. Sometimes variables are just initialized by NULLs and there's no way for error to occur yet. In such cases I put ```if (false) {```. I could move such blocks around to make it look prettier, shove them to the end of function for example, but I think uniform style is preferable.

Granted some functions have only one or two CHECKs, and might be simplified by removing cleanup block altogether. This is because I've been simplifying some functions when coding. I strive to keep only those cleanup blocks that spare me at least some code duplication.

Also I find such blocks much easier to work with when refactoring. Instead of remembering to delete the pointer in this or that case, I just delete them all in cleanup section. And the list of affected pointers is plain to see at the top of the function.

Reply:
btw CHECK macro is what does ```goto fail_cleanup``` when condition is zero